### ISSUE TITLE
Add ifnot

### ISSUE THREAD

**Issue by michaelwinch** [fsharp/fslang-suggestions#1393]
I propose we add an `ifnot` keyword. Arguably it's only useful in simple cases although it will help wean people (myself) off of using backpipes in more places
```
if not <| isSuccessful statusCode then
    do! logError
```
vs
```
ifnot isSuccessful statusCode then
    do! logError
```

## Pros

- Readability
- Reduces use of the backpipe `<|`

## Affidavit (please submit!)

Please tick these items by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [StackOverflow](http://stackoverflow.com)) and I have searched StackOverflow for discussions of this issue
* [x] This is a language change and not purely a tooling change (e.g. compiler bug, editor support, warning/error messages, new warning, non-breaking optimisation) belonging to [the compiler and tooling repository](https://github.com/dotnet/fsharp)
* [x] This is not something which has obviously "already been decided" in previous versions of F#. If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by brianrourkeboll** [comment]
I would put this in the "decided in F# 1.0" (really, decided in OCaml or before) category. You can always just use parentheses :)

**Comment by vzarytovskii** [comment]
+1 to what Brian said, plus it will be a breaking change, since it will be introducing new keyword.

**Comment by edgarfgp** [comment]
A deprecation of the back pipe `<|` would be interesting.

**Comment by brianrourkeboll** [comment]
> How adding a keyword is a breaking change. We recently added. `while!`. ?

`while!` wasn't a valid identifier that someone could've had in their existing code. `ifnot` is currently a valid identifier:

```fsharp
let ifnot = 3
```

So it would need to be a contextual keyword.

**Comment by vzarytovskii** [comment]
> > How adding a keyword is a breaking change. We recently added. `while!`. ?
> 
> `while!` wasn't a valid identifier that someone could've had in their existing code. `ifnot` is currently a valid identifier:
> 
> ```fsharp
> let ifnot = 3
> ```
> 
> So it would need to be a contextual keyword.

`ifnot` is a valid identifier now, `while!` wasn't.
 
 **Update**: Brian beat me to it.

**Comment by michaelwinch** [comment]
> > How adding a keyword is a breaking change. We recently added. `while!`. ?
> 
> `while!` wasn't a valid identifier that someone could've had in their existing code. `ifnot` is currently a valid identifier:
> 
> ```fsharp
> let ifnot = 3
> ```
> 
> So it would need to be a contextual keyword.

Fair point!

**Comment by vzarytovskii** [comment]
> So it would need to be a contextual keyword.

Which, in case of `if` or `ifnot`, shall be parsed in pairs with `then` in the expressions.

**Comment by vzarytovskii** [comment]
Another question, shall there be `elifnot` then? /s

**Comment by edgarfgp** [comment]
@vzarytovskii At the very least we should consider deprecating  back pipes `<| and <||`

**Comment by vzarytovskii** [comment]
> @vzarytovskii At the very least we should consider deprecating back pipes `<| and <||`

Yeah, that would be a separate suggestion. It is also falling under a good use case for analyzers as well.

**Comment by edgarfgp** [comment]
> Another question, shall there be `elifnot` then? /s

I guess not. Before we should add `elif!`

**Comment by edgarfgp** [comment]
> > @vzarytovskii At the very least we should consider deprecating back pipes `<| and <||`
> 
> Yeah, that would be a separate suggestion. It is also falling under a good use case for analyzers as well.

 Im referring compiler deprecation warning like we have done in the past. At this stage not sure if we ever have a proper official F# Analyser support TBH.

**Comment by vzarytovskii** [comment]
That said, I'm going to add `probably not` to it, and close it, as it was decided long time ago. If there's more desire in future, we will reopen it for further disucssion.

**Comment by Martin521** [comment]
> At the very least we should consider deprecating back pipes <|

What's wrong with backpipes in situations like the one in the OP?
I also like it in 
```
lock x <| fun () ->
    // 10 lines of code
```

better than having a closing parenthesis at the end of the 10 lines.

**Comment by vzarytovskii** [comment]
> > At the very least we should consider deprecating back pipes <|
> 
> 
> 
> What's wrong with backpipes in situations like the one in the OP?
> 
> I also like it in 
> 
> ```
> 
> lock x <| fun () ->
> 
>     // 10 lines of code
> 
> ```
> 
> 
> 
> better than having a closing parenthesis at the end of the 10 lines.

Personally - nothing wrong with it, but suggestion is still valid, and can be reviewed and discussed.

**Comment by edgarfgp** [comment]
@Martin521  I agree with Don's statements here `Don't use it, just say NO!` https://youtu.be/1AZA1zoP-II?si=ScKYeYsHREa2cwcT&t=2128 for the specific part.

**Comment by charlesroddie** [comment]
I think an amended version of this is a good idea: allow omission of `then`.

```fsharp
if [condition] else [action]
```

I remember using this a lot when I first learned to program in BASIC. Not sure how common this is in other languges.
